cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java

```diff
@@ -134,23 +159,53 @@ protected final String loadMessageId() {
   @Nullable
   @Override
   protected String processMessages(Iterable<ImmutablePair<String, T>> messages) throws Exception {
-    MessageTrackingIterator iterator;
+    if (this.txSize == this.getFetchSize()) {
+      return doProcess(messages);
+    }
 
-    // Process the notifications and record the message id of where the processing is up to.
-    // 90% of the tx timeout is .9 * 1000 * txTimeoutSeconds = 900 * txTimeoutSeconds
-    long timeBoundMillis = 900L * txTimeoutSeconds;
-    iterator = TransactionRunners.run(getTransactionRunner(), context -> {
-      TimeBoundIterator<ImmutablePair<String, T>> timeBoundMessages = new TimeBoundIterator<>(messages.iterator(),
-                                                                                              timeBoundMillis);
-      MessageTrackingIterator trackingIterator = new MessageTrackingIterator(timeBoundMessages);
-      processMessages(context, trackingIterator);
-      String lastMessageId = trackingIterator.getLastMessageId();
+    TimeBoundIterator timeBoundIterator = new TimeBoundIterator<>(messages.iterator(), timeBoundMillis);
+    MessageTrackingIterator messageTrackingIterator = new MessageTrackingIterator(timeBoundIterator);
+    List<ImmutablePair<String, T>> currentBatch;
+    String lastMessageId = null;
+    while (timeBoundIterator.hasNext()) {
+      currentBatch = new ArrayList<>();
+      if (!messageTrackingIterator.hasNext()) {
+        // message should process in a separate txn
+        messageTrackingIterator = new MessageTrackingIterator(timeBoundIterator);
+      }
+      while (messageTrackingIterator.hasNext() && currentBatch.size() < this.txSize) {
+        currentBatch.add(messageTrackingIterator.next());
```

Comment:
I would recommend to reduce number of wrapping iterators, e.g. Tracking Iterator logic can be done on the batching layer. We don't need to refetch, we only need separate transactions

Reply:
We do need tracking iterator because of TimeBound. The confusing part was because of TrackingIterator doing two things: tracking last message and separating transactions if shouldRunInSeparateTx returned true. 
Separated these two logics into different iterators. Should be simpler now.

Reply:
For completeness on this thread: we ended up removing shouldRunInSeparateTx check from Tracking Iterator, and move it inside the while loop for simplicity.